substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/sources/SourceCache.java

```diff
@@ -0,0 +1,299 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.hosted.image.sources;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.attribute.FileTime;
+import java.util.ArrayList;
+import java.util.List;
+/**
+ * An abstract cache manager for some subspace of the
+ * JDK, GraalVM or application source file space. This class
+ * implements core behaviours that manage a cache of source
+ * files in a specific subdirectory of the local sources
+ * directory. It allows source files to be located
+ * when present in the local cache or cached when not
+ * already present. Subclasses are responsible for providing
+ * behaviours that identify an original source for addition
+ * to the cache and for verifying that a cached file is not
+ * out of date with respect to its original.
+ */
+
+public abstract class SourceCache {
+
+    /*
+     * properties needed to locate relevant JDK and app source roots
+     */
+    protected static final String JAVA_CLASSPATH_PROP = "java.class.path";
+    protected static final String JAVA_HOME_PROP = "java.home";
+    protected static final String JAVA_SPEC_VERSION_PROP = "java.specification.version";
+    /**
+     * A list of root directories which may contain source files
+     * from which this cache can be populated
+     */
+    protected List<Path> srcRoots;
+
+    /**
+     * Create a source cache with a specific base type.
+     * @param key a String identifying the subdir under
+     * which sources should be cached which should also
+     * match the type of content being cached
+     */
+    protected SourceCache(String key) {
+        basePath = Paths.get(SOURCE_CACHE_ROOT_DIR, key);
+        srcRoots = new ArrayList<>();
+
+    }
+
+    /**
+     * A local directory serving as the root for all
+     * source trees maintained by the different
+     * available source caches.
+     */
+    private static final String SOURCE_CACHE_ROOT_DIR = "sources";
+    /**
+     * The top level path relative to the root directory
+     * under which files belonging to this specific cache
+     * are located.
+     */
+    private Path basePath;
+    /**
+     * JDK runtime code sources are cached using this key as a
+     * leading path prefix with a module name as a sub-path
+     * prefix when we have a modular JDK.
+     *
+     * For example, the full file path to a file under the cache
+     * root directory might be jdk/java/lang/Throwable.java on jdk8 or
+     * jdk/java.base/java/lang/Throwable.java on jdk11
+     */
+    protected static final String JDK_CACHE_KEY = "jdk";
```

Comment:
How about
```
public enum SourceCacheType {
    JDK("jdk"),
    GRAALVM("graal"),
    APPLICATION("src");

    final Path subdir;

    SourceCacheType(String subdir) {
        this.subdir = Paths.get(subdir);
    }
}
```
instead of those `protected static final` fields ?

Reply:
Yes, that looks better.

Reply:
I pushed this and used it to clean up the constructors for SourceCache and its subclasses